torch/distributed/fsdp/_optim_utils.py

```diff
@@ -475,12 +458,21 @@ def _flatten_optim_state_dict(
             if flat_state:
                 flat_osd_state[key] = flat_state
             elif optim is not None:  # NamedOptimizer or KeyedOptimizer case.
-                assert len(fqns) == 1
-                local_wrapped_fqn = local_state_clean_fqns.get(fqn, "")
-                if local_wrapped_fqn:
-                    flat_osd_state[key] = copy.deepcopy(
-                        local_state_dict[local_wrapped_fqn]
+                assert len(fqns) == 1 and use_orig_params
+                flat_osd_state[key] = copy.deepcopy(
+                    tree_map_only(
+                        torch.Tensor,
+                        lambda t: t.cpu()
+                        if t.dim() == 0
+                        else torch.zeros(0, device=t.device, dtype=t.dtype),
```

Comment:
Is there any concern that there may be a mixture of CPU and GPU states? (I am guessing that `t.device` might be GPU?)

Reply:
I don't think this is an issue as the step is generally on CPU and other states may be moved to GPU. Moreover, this one is specific to the zero-size states which are not going to be used because the parameters are empty.